cmd/ipfs/daemon.go

```diff
@@ -899,6 +906,61 @@ func serveHTTPGateway(req *cmds.Request, cctx *oldcmds.Context) (<-chan error, e
 	return errc, nil
 }
 
+const gatewayProtocolID protocol.ID = "/ipfs/gateway" // FIXME: specify https://github.com/ipfs/specs/issues/433
+
+func serveTrustlessGatewayOverLibp2p(cctx *oldcmds.Context) (<-chan error, error) {
+	node, err := cctx.ConstructNode()
+	if err != nil {
+		return nil, fmt.Errorf("serveHTTPGatewayOverLibp2p: ConstructNode() failed: %s", err)
+	}
+	cfg, err := node.Repo.Config()
+	if err != nil {
+		return nil, fmt.Errorf("could not read config: %w", err)
+	}
+
+	if !cfg.Experimental.GatewayOverLibp2p {
+		errCh := make(chan error)
+		close(errCh)
+		return errCh, nil
+	}
+
+	opts := []corehttp.ServeOption{
+		corehttp.MetricsCollectionOption("libp2p-gateway"),
+		corehttp.Libp2pGatewayOption(),
+		corehttp.VersionOption(),
+	}
+
+	handler, err := corehttp.MakeHandler(node, nil, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	h := p2phttp.Host{
+		StreamHost: node.PeerHost,
+	}
+
+	tmpProtocol := protocol.ID("/kubo/delete-me")
+	h.SetHTTPHandler(tmpProtocol, http.NotFoundHandler())
+	h.WellKnownHandler.RemoveProtocolMeta(tmpProtocol)
+
+	h.WellKnownHandler.AddProtocolMeta(gatewayProtocolID, p2phttp.ProtocolMeta{Path: "/"})
+	h.ServeMux = http.NewServeMux()
+	h.ServeMux.Handle("/", handler)
```

Comment:
AFAICT mounting at `/` makes testing easier for us (since we can just use the `p2p forward` command) and lowers the complexity of things like redirects. This means we can have a proxy that listens on `localhost:1234` and forwards to the http-over-libp2p endpoint without needing to worry about path rewrites that could be annoying.

I think we should feel free to move this going forward though since the point of `.well-known/libp2p` is to make the mountpoint flexible by selecting on the protocol ID.

Reply:
Note: we have to do the work here a little more explicitly than if we weren't mounting at `/` due to https://github.com/libp2p/go-libp2p/issues/2545